packages/lesswrong/server/callbacks/conversationCallbacks.ts

```diff
@@ -0,0 +1,97 @@
+import { ModeratorActions } from '../../lib/collections/moderatorActions';
+import { FLAGGED_FOR_N_DMS, MAX_ALLOWED_CONTACTS_BEFORE_BLOCK, MAX_ALLOWED_CONTACTS_BEFORE_FLAG } from '../../lib/collections/moderatorActions/schema';
+import { loggerConstructor } from '../../lib/utils/logging';
+import Users from '../../lib/vulcan-users';
+import { getCollectionHooks } from '../mutationCallbacks';
+import { createMutator, updateMutator } from '../vulcan-lib';
+
+/**
+ * Before a user has been fully approved, keep track of how many users they've
+ * messaged.[1] If they've messaged more than N, flag them for review. If
+ * they've messaged more than M, block them from messaging anyone else.
+ *
+ * In the case where a user should be blocked, this will throw an error, so
+ * we should make sure to handle that on the frontend.
+ *
+ * [1] Strictly speaking, this is the number of users with whom they've created
+ * a conversation.
+ */
+async function flagOrBlockUserOnManyDMs({
+  currentConversation,
+  oldConversation,
+  currentUser,
+  context,
+}: {
+  currentConversation: Partial<DbConversation>,
+  oldConversation?: DbConversation,
+  currentUser: DbUser|null,
+  context: ResolverContext,
+}): Promise<Partial<DbConversation>> {
+  const logger = loggerConstructor('callbacks-conversations');
+  logger('flagOrBlockUserOnManyDMs()')
+  if (!currentUser) {
+    throw new Error("You can't create a conversation without being logged in");
+  }
+  if (currentUser.reviewedByUserId && !currentUser.snoozedUntilContentCount) {
+    logger('User has been fully approved, ignoring')
+    return currentConversation;
+  }
+  // if the participants didn't change, we can ignore it
+  if (!currentConversation.participantIds) {
+    logger('No change to participantIds, ignoring')
+    return currentConversation;
+  }
+
+  // Old conversation *should* be completely redundant with
+  // currentUser.usersContactedBeforeReview, but we will try to be robust to
+  // the case where it's not
+  logger('previous usersContactedBeforeReview', currentUser.usersContactedBeforeReview)
+  const allUsersEverContacted = [...(new Set([
+    ...currentConversation.participantIds,
+    ...(oldConversation?.participantIds ?? []),
+    ...(currentUser.usersContactedBeforeReview ?? [])
+  ]))].filter(id => id !== currentUser._id)
+  logger(
+    'new allUsersEverContacted', allUsersEverContacted,
+    '(length: ', allUsersEverContacted.length, ')'
+  )
+  if (allUsersEverContacted.length > MAX_ALLOWED_CONTACTS_BEFORE_FLAG) {
+    // Flag the user
+    logger('Flagging user')
+    void createMutator({
+      collection: ModeratorActions,
+      context,
+      currentUser: null,
+      validate: false,
+      document: {
+        userId: currentUser._id,
+        type: FLAGGED_FOR_N_DMS,
+      },
+    });
+  }
+  // Always update the numUsersContacted field, for denormalization
+  void updateMutator({
```

Comment:
I don't love `void`-ing these two promises, but maybe I'm being over-paranoid

Reply:
What's the concern? I would hope our mutator system is robust to two mutators changing different fields on the same user. I've spent a lot of time in this part of the code recently and I can't see any reason why not.

In general I think void is a pretty valid tool?

Reply:
I just get a bit worried that if you hypothetically `await`ed a call to this function, then we'd return the to caller in an incomplete state. The function isn't exported though so it should be fine here.

It looks like there's actually no `await` in this function, so it doesn't need to be `async`. That's getting a little nit-picky though.

Basically, I think this is fine to merge - I'm just being difficult 🙃 .